### PR TITLE
Develop Database

### PR DESCRIPTION
**Pull request by sunju0123**
Change 3 files

- DataBases
- DataBaseOpenHelper
- ScriptedURL

**DataBases**
update database table, add `TITLE`, `rep image url`, `content` column
I checked @leesera 's PR, then I find expectedTime variable. 
If we want using double data type in sqlite, will change integer type -> **real type**

**DataBaseOpenHelper**
update `column constant`
update `onUpgrade` method. If change the database version, this method will run. Then don't remove application when database update.
update `getUrlListFromQuery` Method. (get all column from database)
add `setTitle`, `setRepImageUrl`, `setContent` Method.
add `updateDataQuery` method. (Because I want remove repeated code. Update some item into specific url)
add `isDuplicationUrl` method. (I want remove repeated code, too. It return true or false. `insertScriptedDataWithCheckDuplication` and `insertScriptedUrl` using it.)
add `insertScriptedUrl` method and `insertScriptedUrlQuery` method. (In V2, we using ScriptedUrl class.)

**ScriptedURL**
add  
```
  private String title;
  private String content;
  private String repImageUrl;
```
and   add 
`ScriptedURL(String url, int isRead, String title, String content, String repImageUrl, int wordCount) 
`
In ScriptedURL I think conflict in word count and expectedTime
So will talk about this issue

### CODE REVIEW

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java [hunk 1]
```diff
@@ -100,6 +122,20 @@ public void deleteAllPage() {
...
+  private boolean isDuplicationUrl(String url) {
+    SQLiteDatabase dataBase = getReadableDatabase();
+    int urlDuplicationCount;
+    Cursor cursor = dataBase.rawQuery("SELECT * FROM PAGE WHERE URL = ('" + url + "') ", null);
+    urlDuplicationCount = cursor.getCount();
+    dataBase.close();
+
+    if (urlDuplicationCount == 0) {
+      return false;
+    } else {
```
Comment (xhae) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: `if () { return false; } return true;`

File: Roys/app/src/main/java/com/dwg_karrier/roys/MainActivity.java [hunk 2]
```diff
@@ -128,8 +128,13 @@ protected String doInBackground(String... params) {
         //dataBaseOpenHelper.deleteAllPage();
         //dataBaseOpenHelper.getTableAsString();
 
+        // TODO: add another check url duplication method. (Without database query.)
+
         for (int i = 0; i < len; i++) {
-          dataBaseOpenHelper.insertScriptedDataWithCheckDuplication(arr.getJSONObject(i).getString("originId"));
+          String urlItem = arr.getJSONObject(i).getString("originId");
+          if (!dataBaseOpenHelper.isDuplicationUrl(urlItem)) {
```
Comment (sunju0123) [inline Roys/app/src/main/java/com/dwg_karrier/roys/MainActivity.java]: add check url duplication before insert url.
change insertScriptedDataWithCheckDuplication ->  insertScriptedData

**Comment by sunju0123** [comment]
add `getScriptedUrlListByTime` method from @Csoyee 's request.

This method returns ScriptedUrl items. Items are `Expectedtime` is over then minTime and less then maxTime item.

**getScriptedUrlListByTime**
   * @param minTime expected time's minimum value
   * @param maxTime expected time's maximum value
   * @return filtered ScriptedUrl List

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java [hunk 3]
```diff
@@ -103,12 +134,22 @@ public void deleteAllPage() {
...
+  /**
+   * check url duplication from database data.
+   * @param url
+   * @return url duplication result. duplication -> true, not duplication -> false
+   */
+  public boolean isDuplicationUrl(String url) {
+    SQLiteDatabase dataBase = getReadableDatabase();
+    int urlDuplicationCount;
+    Cursor cursor = dataBase.rawQuery("SELECT * FROM PAGE WHERE URL = ('" + url + "') ", null);
+    urlDuplicationCount = cursor.getCount();
```
Comment (xhae) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: int urlDuplicationCOunt = cursor.getCount();

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java [hunk 4]
```diff
@@ -103,12 +134,22 @@ public void deleteAllPage() {
...
-    SQLiteDatabase dataBase = getWritableDatabase();
-    String escapedTitle = StringEscapeUtils.escapeHtml4(title);
-    String escapedContent = StringEscapeUtils.escapeHtml4(content);
-    dataBase.execSQL("INSERT INTO PAGE (URL, TITLE, CONTENT, EXPECTEDTIME, IMGURL) VALUES ('" + url + "',\"" + escapedTitle + "\", \"" + escapedContent + "\", " + String.valueOf(expectedTime) + " , '" + imgUrl + "');");
+  /**
+   * check url duplication from database data.
+   * @param url
+   * @return url duplication result. duplication -> true, not duplication -> false
+   */
+  public boolean isDuplicationUrl(String url) {
```
Comment (xhae) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: In the semantic view, I recommend to use `isDuplicatedUrl` instead of `Duplication`.

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java (hunk 3 above)
Comment (sunju0123) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: Done!

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java (hunk 4 above)
Comment (sunju0123) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: Done!

**Comment by sunju0123** [comment]
Modified some line

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java [hunk 5]
```diff
@@ -103,12 +134,21 @@ public void deleteAllPage() {
...
-    SQLiteDatabase dataBase = getWritableDatabase();
-    String escapedTitle = StringEscapeUtils.escapeHtml4(title);
-    String escapedContent = StringEscapeUtils.escapeHtml4(content);
-    dataBase.execSQL("INSERT INTO PAGE (URL, TITLE, CONTENT, EXPECTEDTIME, IMGURL) VALUES ('" + url + "',\"" + escapedTitle + "\", \"" + escapedContent + "\", " + String.valueOf(expectedTime) + " , '" + imgUrl + "');");
+  /**
+   * check url duplication from database data.
+   * @param url
+   * @return url duplication result. duplication -> true, not duplication -> false
+   */
+  public boolean Duplication(String url) {
```
Comment (xhae) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: Please read the comment again.

File: Roys/app/src/main/java/com/dwg_karrier/roys/ScriptedURL.java [hunk 6]
```diff
@@ -54,6 +55,20 @@
...
+    this.url = url;
+    if (isRead == 1) {
+      this.isRead = true;
+    } else {
+      this.isRead = false;
+    }
+    this.title = title;
+    this.content = content;
+    this.repImageUrl = repImageUrl;
+    this.wordCount = wordCount;
```
Comment (sunju0123) [inline Roys/app/src/main/java/com/dwg_karrier/roys/ScriptedURL.java]: I noticed about that line. But no conflict issue. So I thought lightly that issue.
OMG My old code and new code were merged...
I'll update code now.

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java [hunk 7]
```diff
@@ -63,36 +66,64 @@ public int getReadPageCount() {
...
   }
 
+  /**
+   * It returns ScriptedUrl List by user's spear time from Database.
+   * @param minTime expected time's minimum value
+   * @param maxTime expected time's maximum value
+   * @return ScriptedUrl List, Item's expected time : over then minTime and less then maxTime.
+   */
+  public ArrayList<ScriptedURL> getScriptedUrlListByTime(int minTime, int maxTime) {
+    String getUrlListQuery = "SELECT * from page where expectedtime >= " + minTime + "and exexpectedtime <= " + maxTime;
```
Comment (sunju0123) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: thank you!

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java [hunk 8]
```diff
@@ -100,12 +133,22 @@ public void deleteAllPage() {
...
+   * @param url
+   * @return checked result. If the url is duplicated , returns true. Else return false.
+   */
+  public boolean isDuplicationUrl(String url) {
+    SQLiteDatabase dataBase = getReadableDatabase();
+    Cursor cursor = dataBase.rawQuery("SELECT * FROM PAGE WHERE URL = ('" + url + "') ", null);
+    int urlDuplicationCount = cursor.getCount();
     dataBase.close();
+
+    if (urlDuplicationCount == 0) {
```
Comment (nuinuri) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: return urlducplicationCount != 0;

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java [hunk 9]
```diff
@@ -128,13 +171,36 @@ public void getTableAsString() {
...
 
-    if (cursor.getCount() == 0) {
-      insertScriptedData(url, title, content, expectedTime, imgUrl);
+  /**
+   * Please Check URL duplication before using insert method.
+   * Using Duplication()
+   * @param scriptedURL
+   */
+  public void insertScriptedUrl(ScriptedURL scriptedURL) {
+    int readValue;
```
Comment (nuinuri) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: int readValue = striptedURL.getIsRead() ? 1 : 0;

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java [hunk 10]
```diff
@@ -128,13 +171,36 @@ public void getTableAsString() {
...
+    if (scriptedURL.getIsRead()) {
+      readValue = 1;
+    } else {
+      readValue = 0;
     }
+    SQLiteDatabase dataBase = getWritableDatabase();
+    dataBase.execSQL("INSERT INTO PAGE (READ, URL, TITLE, repImage, CONTENT, WORDCOUNT) VALUES ("
```
Comment (nuinuri) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: String.format("xxxxxx", ddd) looks better.

File: Roys/app/src/main/java/com/dwg_karrier/roys/ScriptedURL.java [hunk 11]
```diff
@@ -54,6 +55,20 @@
     this.crawler = new Crawler(url);
   }
 
+  ScriptedURL(String url, int isRead, String title, String content, String repImageUrl, double expectedTime) {
+    this.url = url;
+    if (isRead == 1) {
```
Comment (nuinuri) [inline Roys/app/src/main/java/com/dwg_karrier/roys/ScriptedURL.java]: this.isRead = isRead == 1;

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java [hunk 12]
```diff
@@ -128,13 +171,36 @@ public void getTableAsString() {
...
+   * Please Check URL duplication before using insert method.
+   * Using Duplication()
+   * @param url
+   */
+  public void insertScriptedData(String url, String title, String content, int expectedTime, String imgUrl) {
+    SQLiteDatabase dataBase = getWritableDatabase();
+    String escapedTitle = StringEscapeUtils.escapeHtml4(title);
+    String escapedContent = StringEscapeUtils.escapeHtml4(content);
+    dataBase.execSQL("INSERT INTO PAGE (URL, TITLE, CONTENT, EXPECTEDTIME, IMGURL) VALUES ('" + url + "',\"" + escapedTitle + "\", \"" + escapedContent + "\", " + String.valueOf((int)expectedTime) + " , '" + imgUrl + "');");
+    dataBase.close();
```
Comment (nuinuri) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: Not sure but. 

try (SQLiteDatabase dataBase = getWriteableDatabase()) {
   xxx. 
}

You don't need to call "close()"

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java [hunk 13]
```diff
@@ -100,12 +133,22 @@ public void deleteAllPage() {
...
+   * check url duplication from database data.
+   * If result returns true, you can't insert url data into DB.
+   * @param url
+   * @return checked result. If the url is duplicated , returns true. Else return false.
+   */
+  public boolean isDuplicationUrl(String url) {
+    SQLiteDatabase dataBase = getReadableDatabase();
+    Cursor cursor = dataBase.rawQuery("SELECT * FROM PAGE WHERE URL = ('" + url + "') ", null);
+    int urlDuplicationCount = cursor.getCount();
     dataBase.close();
```
Comment (nuinuri) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: Ditto. Would you apply autocloseable in java feature?

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java [hunk 14]
```diff
@@ -78,18 +93,36 @@ public int getReadPageCount() {
...
+
+  public void setContent(String url, String content) {
+    String query = "content = '" + content + "'";
+    updateDataQuery(query, url);
+  }
+
+  private void updateDataQuery(String query, String url) {
+    SQLiteDatabase dataBase = getWritableDatabase();
+    dataBase.execSQL("UPDATE PAGE SET " + query + " WHERE url='" + url + "';");
     dataBase.close();
```
Comment (nuinuri) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: try (SQLiteDatabase dataBase = getWritableDatabase()) {
  dataBase.execSQL();
}

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java [hunk 15]
```diff
@@ -63,12 +66,24 @@ public int getReadPageCount() {
...
+    return getUrlListFromQuery(getUrlListQuery);
+  }
+
   private ArrayList<ScriptedURL> getUrlListFromQuery(String query) {
     SQLiteDatabase dataBase = getReadableDatabase();
     ArrayList<ScriptedURL> resultList = new ArrayList<ScriptedURL>();
 
     Cursor cursor = dataBase.rawQuery(query, null);
     while (cursor.moveToNext()) {
+
```
Comment (nuinuri) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: Unnecessary whitespace

File: Roys/app/src/main/java/com/dwg_karrier/roys/ScriptedURL.java (hunk 11 above)
Comment (sunju0123) [inline Roys/app/src/main/java/com/dwg_karrier/roys/ScriptedURL.java]: #42

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java (hunk 10 above)
Comment (sunju0123) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: #42

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java (hunk 12 above)
Comment (sunju0123) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: #42

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java (hunk 13 above)
Comment (sunju0123) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: #42

File: Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java (hunk 14 above)
Comment (sunju0123) [inline Roys/app/src/main/java/com/dwg_karrier/roys/DataBaseOpenHelper.java]: #42
